File: 프로그래머스/1/172928. 공원 산책/공원 산책.cpp
#include <string>
#include <vector>

using namespace std;

vector<int> solution(vector<string> park, vector<string> routes) {
    vector<int> answer; 
    pair<int, int> start;
    
    for(int i=0; i<park.size(); i++){
        for(int j=0; j<park[i].length(); j++){
            if(park[i][j] == 'S'){
                start = make_pair(i, j);
                break;
            }
        }
    }
    
    for(int i=0; i<routes.size(); i++){
        pair<int, int> tmp = start;
        bool check = false;
        int n = stoi(routes[i].substr(routes[i].find(' ')+1, routes[i].length()));
        
        if(routes[i][0] == 'E'){
            for(int j=0; j<n; j++){
                if(start.second == park[0].length()-1 || park[start.first][start.second+1] == 'X'){
                    check = true;
                    break;
                }
                
                start.second++;
            }
        }
        if(routes[i][0] == 'W'){
            for(int j=0; j<n; j++){
                if(start.second == 0 || park[start.first][start.second-1] == 'X'){
                    check = true;
                    break;
                }
                
                start.second--;
            }
        }
        if(routes[i][0] == 'N'){
            for(int j=0; j<n; j++){
                if(start.first == 0 || park[start.first-1][start.second] == 'X'){
                    check = true;
                    break;
                }
                
                start.first--;
            }
        }
        if(routes[i][0] == 'S'){
            for(int j=0; j<n; j++){
                if(start.first == park.size()-1 || park[start.first+1][start.second] == 'X'){
                    check = true;
                    break;
                }
                
                start.first++;
            }
        }
        
        if(check)
            start=tmp;
    }
    
    answer.push_back(start.first);
    answer.push_back(start.second);
    
    return answer;
}

File: 프로그래머스/1/172928. 공원 산책/공원 산책.js
function solution(park, routes) {
    var answer = [];
    
    let height, width;
    
    for(let i=0; i<park.length; i++){
        for(let j=0; j<park[i].length; j++){
            if(park[i][j] === 'S')
                height = i, width = j;
        }
    }
    
    for(let i=0; i<routes.length; i++){
        let n = parseInt(routes[i].substring(routes[i].indexOf(' ')+1));
        let check = false;
        let tmpH = height, tmpW = width;

        if(routes[i][0] === 'E'){
           for(let j=0; j<n; j++){
               if(width === park[0].length-1 || park[height][width+1] === 'X'){
                   check = true;
                   break;
               }
               
               width++;
           }
        }
        if(routes[i][0] === 'W'){
           for(let j=0; j<n; j++){
               if(width === 0 || park[height][width-1] === 'X'){
                   check = true;
                   break;
               }
               
               width--;
           }
        }
        if(routes[i][0] === 'S'){
           for(let j=0; j<n; j++){
               if(height === park.length-1 || park[height+1][width] === 'X'){
                   check = true;
                   break;
               }
               
               height++;
           }
        }
        if(routes[i][0] === 'N'){
           for(let j=0; j<n; j++){
               if(height === 0 || park[height-1][width] === 'X'){
                   check = true;
                   break;
               }
               
               height--;
           }
        }
        
        if(check === true)
            height = tmpH, width = tmpW;
    }
    
    answer.push(height);
    answer.push(width);

    return answer;
}